### PR TITLE
DBZ-2834 Make outbox extensions dependency on tracing extension optional

### PR DESCRIPTION
**Pull request by ani-sha**
https://issues.redhat.com/browse/DBZ-2834

### CODE REVIEW

**Comment by jpechane** [comment]
Could you please check the changes in https://github.com/debezium/debezium/pull/2407 and verify if they needs to be included here too?
@Naros Do you think it is better to merge the Quarkus upgrade PR with this one and make the changes in sync?

**Comment by ani-sha** [comment]
@jpechane  yes the changes need to be included here. Should we add the changes here and merge this first, so that the other PR can wait for the Quarkus 2.0 release.

**Comment by jpechane** [comment]
@ani-sha I've discussed it with @gunnarmorling and the conlusion is the other way around. Let's upgrade Quarkus first and then you'll rebase this PR of it.

File: documentation/modules/ROOT/pages/integrations/tracing.adoc [hunk 1]
```diff
@@ -96,6 +96,12 @@ When an outbox even is emitted the extension will:
 * inject the span context into that `java.util.Properties` instance that is serialized into the `tracingspancontext` column
 * write the record into the database
 
+[NOTE]
+====
+The `smallrye-opentracing` dependency is an optional dependency in the Outbox extension. It can be configured by setting an option `quarkus.debezium-outbox.tracing.enabled=true`
+in the Quarkus `application.properties` file.
```
Comment (gunnarmorling) [inline documentation/modules/ROOT/pages/integrations/tracing.adoc:102]: I'm not sure whether I get this sentence right, but by default, the sole presence of the OpenTracing extension in the build should enable the tracing usage in the Outbox extension. I.e. no explicit enabled=true setting should be required for that. Only to *disable* tracing despite the OpenTracing extension being present should require an explicit setting. You should find precedence for this in the Quarkus Kafka Streams extension (which does the same for the health extension); alternatively, @Naros or the folks from Quarkus community should be able to help.

**Review by gunnarmorling — COMMENTED** [review]
One question inline.

File: documentation/modules/ROOT/pages/integrations/tracing.adoc (hunk 1 above)
Comment (Naros) [reply documentation/modules/ROOT/pages/integrations/tracing.adoc:102]: So @ani-sha I think all we need to change is the default value for the `quarkus.debezium-outbox.tracing.enabled` config option should be _true_ and then we should adjust the documentation based on Gunnar's comments.  This should align with what Gunnar is asking I believe.

File: debezium-quarkus-outbox/deployment/src/main/java/io/debezium/outbox/quarkus/deployment/OutboxProcessor.java [hunk 2]
```diff
@@ -141,8 +144,16 @@ public void produceOutboxBuildItem(CombinedIndexBuildItem index,
...
                       BuildProducer<AdditionalBeanBuildItem> additionalBeanProducer,
                       BuildProducer<GeneratedResourceBuildItem> generatedResourcesProducer,
-                      BuildProducer<ReflectiveClassBuildItem> reflectiveClassProducer) {
-        additionalBeanProducer.produce(AdditionalBeanBuildItem.unremovableOf(EventDispatcher.class));
+                      BuildProducer<ReflectiveClassBuildItem> reflectiveClassProducer,
+                      BuildProducer<OutboxOpenTracingBuildItem> buildItemBuildProducer,
+                      Capabilities capabilities) {
+        if (debeziumOutboxConfig.tracingEnabled && capabilities.isPresent(Capability.OPENTRACING)) {
+            additionalBeanProducer.produce(AdditionalBeanBuildItem.unremovableOf(DebeziumTracerEventDispatcher.class));
+            buildItemBuildProducer.produce(new OutboxOpenTracingBuildItem(true));
```
Comment (gunnarmorling) [inline debezium-quarkus-outbox/deployment/src/main/java/io/debezium/outbox/quarkus/deployment/OutboxProcessor.java:152]: Where is this build item used later on? Is it actually needed?

File: debezium-quarkus-outbox/integration-tests/src/test/java/io/debezium/outbox/quarkus/it/OutboxProfiles.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+public class OutboxProfiles {
+    /**
+     * Uses the values directly from {@code application.properties} without changes.
+     */
+    public static class Default implements QuarkusTestProfile {
+    }
+
+    /**
+     * Enables OpenTracing support that is disabled by default.
+     */
```
Comment (gunnarmorling) [inline debezium-quarkus-outbox/integration-tests/src/test/java/io/debezium/outbox/quarkus/it/OutboxProfiles.java:27]: I'm still a bit confused by this. IMO, The rules should be:

* If the OT extension itself is not present, (naturally), there's no tracing support provided by the outbox extension
* If the OT extension itself is present, the outbox tracing support is enabled by default
* If the OT extension itself is present, the outbox tracing support can optionally be disabled by setting `quarkus.debezium-outbox.tracing.enabled` to `false`; 

Is this what we're doing?

File: debezium-quarkus-outbox/deployment/src/main/java/io/debezium/outbox/quarkus/deployment/DebeziumOutboxConfig.java [hunk 4]
```diff
@@ -70,6 +70,12 @@
     @ConfigItem
     public DebeziumOutboxConfigTracingSpan tracingSpan;
 
+    /**
+     * Whether or not smallrye-opentracing extension is present.
+     */
+    @ConfigItem(name = "tracing.enabled", defaultValue = "false")
```
Comment (gunnarmorling) [inline debezium-quarkus-outbox/deployment/src/main/java/io/debezium/outbox/quarkus/deployment/DebeziumOutboxConfig.java:76]: This should default to true, as discussed elsewhere. Also the comment isn't quite right.

File: debezium-quarkus-outbox/integration-tests/src/test/java/io/debezium/outbox/quarkus/it/OutboxOpenTracingTest.java [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+import io.quarkus.test.junit.TestProfile;
+
+/**
+ * Integration tests for the Debezium Outbox extension, enabling OpenTracing behavior by default.
+ *
+ * @author Chris Cranford
+ */
+@QuarkusTest
+@TestProfile(OutboxProfiles.OpenTracing.class)
+public class OutboxOpenTracingTest extends AbstractOutboxTest {
```
Comment (gunnarmorling) [inline debezium-quarkus-outbox/integration-tests/src/test/java/io/debezium/outbox/quarkus/it/OutboxOpenTracingTest.java:18]: How is the tracing functionality being tested? Validating a non-null value in the trace id column seems like a reasonable approach?

**Review by gunnarmorling — COMMENTED** [review]
@ani-sha, @Naros, I think we're getting there. Some last (hopefully anyways ;) comments inline.

File: debezium-quarkus-outbox/integration-tests/src/test/java/io/debezium/outbox/quarkus/it/OutboxProfiles.java (hunk 3 above)
Comment (ani-sha) [reply debezium-quarkus-outbox/integration-tests/src/test/java/io/debezium/outbox/quarkus/it/OutboxProfiles.java:27]: @gunnarmorling In the confusion, we have done the opposite to the rules said ie. OT is disabled by default and will be enabled via `quarkus.debezium-outbox.tracing.enabled` as `true`.

**Review by gunnarmorling — COMMENTED** [review]
Pushed one more commit for some clean-up. Merging now. Thanks a lot, @ani-sha and @Naros, great job!

**Pull request merged**